pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -80,62 +75,18 @@ func (configgen *ConfigGeneratorImpl) BuildClusters(proxy *model.Proxy, req *mod
 	return configgen.buildClusters(proxy, req, services)
 }
 
-// BuildDeltaClusters generates the deltas (add and delete) for a given proxy. Currently, only service changes are reflected with deltas.
-// Otherwise, we fall back onto generating everything.
-func (configgen *ConfigGeneratorImpl) BuildDeltaClusters(proxy *model.Proxy, updates *model.PushRequest,
+// BuildDeltaClusters generates the deltas (add and delete) for a given proxy.
+func (configgen *ConfigGeneratorImpl) BuildDeltaClusters(proxy *model.Proxy, req *model.PushRequest,
```

Comment:
Couple of questions:
- So in Delta we generate all the resources as usual and later filter based on cache token? Then why do not we call regular BuildClusters rather than repeating the logic here?
- Do we still need, usedDelta?

Reply:
> So in Delta we generate all the resources as usual and later filter based on cache token? Then why do not we call regular BuildClusters rather than repeating the logic here?

If this is the case (didn't read too much), that would be not great. Since control plane load will still be just as high as without delta mostly.

Can we just construct the key (cheap) and see if they key depends on the recent push?

Reply:
I agree. That was my next questions if the answer to my first question is "Yes". It is not definetly not great, we are only optimizing on network (with additional overhead of filtering in control plane) if I read the code correctly?

Reply:
>So in Delta we generate all the resources as usual and later filter based on cache token? Then why do not we call regular BuildClusters rather than repeating the logic here?

We can not use `BuildClusters without any change, but can refactor it to return the token info. Then we can decide the delta changes.

>Can we just construct the key (cheap) and see if they key depends on the recent push?

I did not thinking too much about this, worth looking into. 

>If this is the case (didn't read too much), that would be not great. Since control plane load will still be just as high as without delta mostly.

Yeah, currently it only saves bandwidth between pilot and sidecar, which is still great from our performance test last year.  

And also helps a lot for envoy config hot reloading.

Reply:
>>Can we just construct the key (cheap) and see if they key depends on the recent push?

This will introduce another cache, for CDS the key is cluster name, the value is the clusterkey. Then is it easy to get the updated clusternames by comparing the clusterkeys.

How about deleted clusters? How could we know?

Reply:
> Yeah, currently it only saves bandwidth between pilot and sidecar, which is still great from our performance test last year.

While it still great, what is our longer term plan? Our longer term plan is to be more "delta" aware so that we optimize on controlplane also along with network optimization. So in that sense, taking this approach now - seems not a step in our long term direction or do you think we can go back to long term approach easily?

Reply:
Why do you think depending on the cache is not ` more "delta" aware`? Or what do you think is `  more "delta" aware`?

Come back to the first question: `So in Delta we generate all the resources as usual and later filter based on cache token? `

When we generate all resources, if one need to regenerate, it absolutely means it is updated. So for it, no matter what kind of way we use, the effort is same.

For one in the cache, I means it has no change at all, and since it is in the cache, why it is heavy?

Reply:
> Why do you think depending on the cache is not more "delta" aware? Or what do you think is more "delta" aware?

I am not saying using cache is not "delta" aware. All I am saying is irrespective of mechanism we choose
- We have to optimize compute in control plane as well, we can not just say it will be OK if we optimize on just network.

Based on what you described above it seems there is no additional overhead of Building all clusters and filtering them for delta in controlplane which is different from your original statement that we are only getting network optimization - That confused me.

Reply:
>Based on what you described above it seems there is no additional overhead of Building all clusters and filtering them for delta in controlplane which is different from your original statement that we are only getting network optimization - That confused me.

sorry to make you confused, the truth is with cache, cpu/badwidth can both be saved for delta xds.